docs/jmespath-grammar.txt

```diff
@@ -0,0 +1,133 @@
+expr      = ws ( terminal / non-terminal ) ws
+root-expr = non-test / not / comparison
+non-test  = subexpr / index / flatten / wildcard-index / filter /
+            identifier / current-node / literal / root-multi-list /
+            multi-hash / function / group / slice
+group     = begin-group expr end-group
+
+; Insignificant whitespace is allowed around expr nodes and around the
+; following structural tokens.
+begin-array     = ws %x5B ws    ; [ left square bracket
+begin-object    = ws %x7B ws    ; { left curly bracket
+end-array       = ws %x5D ws    ; ] right square bracket
+end-object      = ws %x7D ws    ; } right curly bracket
+begin-group     = ws %x28 ws    ; ( left parenthesis
+end-group       = ws %x29 ws    ; ) right parenthesis
+name-separator  = ws %x3A ws    ; : colon
+value-separator = ws %x2C ws    ; , comma
+decimal-point   = ws %x2E ws    ; .
+pipe-separator  = ws "|" ws
+or-separator    = ws "||" ws
+and-separator   = ws "&&" ws
+lt              = ws "<" ws
+lte             = ws "<=" ws
+gt              = ws ">" ws
+gte             = ws ">=" ws
+eq              = ws "==" ws
+ne              = ws "!=" ws
+expref-token    = "&" ws
+not-token       = "!" ws
+begin-filter    = "[?" ws
+flatten         = ws "[]" ws
+
+; Insignificant whitespace
+ws = *(%x20 / ; Space
+       %x09 / ; Horizontal tab
+       %x0A / ; Line feed or New line
+       %x0D) ; Carriage return
+
+; "&&" binds more tightly than "||"
+; "||" binds more tightly than "|".
+terminal     = pipe / or / and
```

Comment:
Just curious, why is this called terminal?  I'm thinking of [terminals](http://en.wikipedia.org/wiki/Terminal_and_nonterminal_symbols) as the leaf nodes in the parsed tree, i.e it can't be replaced with anything.  But everything on the RHS can be broken down into more nodes.